lib/game.rb

```diff
@@ -36,114 +36,127 @@ def setup_game
-    #   while !@player_board.place(ship, gets.chomp.upcase.split)
-    #     @message.ship_placement_invalid_coordinates
-    #   end
-    #   puts @player_board.render(true)
-    #  }
+    @player_board.place_user_ships
 
      play
   end
 
   def select_difficulty_level
-    @message.choose_difficulty
+    Messages.choose_difficulty
     player_choice = gets.chomp
-    if ['e','h'].include?(player_choice)
-      return player_choice
-    else
-      @message.invalid_input
-      return select_difficulty_level
+    until ['E','H'].include?(player_choice.upcase)
+      Messages.invalid_input
+      player_choice = gets.chomp
     end
-  end
+    return player_choice
 
-  def select_ship_attributes(board_area, rows, columns)
+  end
 
-    @message.would_you_like_to_customize_ships?
+  def select_ship_attributes( rows, columns)
+    board_area = rows * columns
+    Messages.would_you_like_to_customize_ships?
     valid_selections = [1, 2, 3, 4]
     ship_selection = gets.chomp.to_i
 
     until valid_selections.include?(ship_selection)
-      @message.invalid_customize_ship_selection
+      Messages.invalid_input
       ship_selection = gets.chomp.to_i
     end
 
     # Ship lengths must be < 1/3 of the board area to play
     case ship_selection
-    when 1
-      if (board_area / 3) < 5
-        @message.board_is_too_small_for_ship_selection
-        return select_ship_attributes(board_area, rows, columns)
-      end
+    when 1 # Smallest possible board allows these ships
       return [['Cruiser', 3],['Submarine', 2]]
+
     when 2
       if (board_area / 3) < 9
-        @message.board_is_too_small_for_ship_selection
-        return select_ship_attributes(board_area, rows, columns)
+        Messages.board_is_too_small_for_ship_selection
+        return select_ship_attributes( rows, columns)
       end
       return [['Battleship', 4],['Cruiser', 3],['Submarine', 2]]
+
     when 3
       if (board_area / 3) < 14
-        @message.board_is_too_small_for_ship_selection
-        return select_ship_attributes(board_area, rows, columns)
+        Messages.board_is_too_small_for_ship_selection
+        return select_ship_attributes(rows, columns)
       end
       return [['Carrier', 5],['Battleship', 4],['Cruiser', 3],['Submarine', 2]]
+
     when 4
-      return user_custom_ships(board_area, rows, columns)
+      return user_custom_ships(rows, columns)
     end
-
   end
 
-  def user_custom_ships(board_area, rows, columns)
+  def user_custom_ships(rows, columns)
+    potential_ship_length = (rows * columns)/3
     user_input = "Y"
     user_ships = []
-    user_ships_total_length = user_ships.sum{ |ship| ship[1]}
+    remaining_ship_length = potential_ship_length
+    # Loop until user_input "N"
+    until user_input == "N"
+
+
+      Messages.prompt_user_for_custom_ship_name(user_ships, remaining_ship_length)
+      ship_name  = gets.chomp
+      until ship_name != "" && !user_ships.map{|ship| ship[0]}.include?(ship_name)
```

Comment:
Nice check for duplicate ship names.  While technically not breaking anything, it was weird when 2 ships are named the same.